lib/Dialect/FIRRTL/Transforms/ExtractClasses.cpp

```diff
@@ -0,0 +1,146 @@
+    return;
+
+  OpBuilder builder = OpBuilder::atBlockEnd(getOperation().getBody(0));
+
+  // Collect the parameter names from input properties.
+  SmallVector<StringRef> formalParamNames;
+  for (auto inputProperty : inputProperties)
+    formalParamNames.push_back(inputProperty.name);
+
+  // Construct the ClassOp with the FModuleOp name and parameter names.
+  auto classOp = builder.create<ClassOp>(moduleOp.getLoc(), moduleOp.getName(),
+                                         formalParamNames);
+
+  // Construct the ClassOp body with block arguments for each input property,
+  // updating the mapping to map from the input property to the block argument.
+  Block *classBody = &classOp.getRegion().emplaceBlock();
+  for (auto inputProperty : inputProperties) {
+    BlockArgument parameterValue =
+        classBody->addArgument(inputProperty.type, inputProperty.loc);
+    BlockArgument inputValue = moduleOp.getArgument(inputProperty.index);
+    mapping.map(inputValue, parameterValue);
+  }
+
+  // Construct ClassFieldOps for each output property.
+  builder.setInsertionPointToStart(classBody);
+  for (auto outputProperty : outputProperties) {
+    // Get the Value driven to the property to use for this ClassFieldOp.
+    auto outputValue =
+        cast<FIRRTLPropertyValue>(moduleOp.getArgument(outputProperty.index));
+    Value originalValue = getDriverFromConnect(outputValue);
+
+    // Mark the property assign to be erased.
+    opsToErase.push_back(getPropertyAssignment(outputValue));
+
+    // If the Value is defined by an Operation, copy that into the body, and
+    // map from the old Value to the new Value. This may need to walk property
+    // ops in order to copy them into the ClassOp, but for now only constant
+    // ops exist. Mark the property op to be erased.
+    if (auto *op = originalValue.getDefiningOp()) {
+      builder.clone(*op, mapping);
+      opsToErase.push_back(op);
+    }
+
+    // Create the ClassFieldOp using the mapping to find the appropriate Value.
+    Value fieldValue = mapping.lookup(originalValue);
+    builder.create<ClassFieldOp>(originalValue.getLoc(), outputProperty.name,
+                                 fieldValue);
+  }
+
+  // Clean up the FModuleOp by removing property ports and operations. This
+  // first erases opsToErase in the order they were added, so property
+  // assignments are erased before value defining ops. Then it erases ports.
+  for (auto *op : opsToErase)
+    op->erase();
+  moduleOp.erasePorts(portsToErase);
+}
+
+/// Extract OM classes from FIRRTL modules with properties.
+void ExtractClassesPass::runOnOperation() {
+  // Get the CircuitOp.
+  auto circuits = getOperation().getOps<CircuitOp>();
+  if (circuits.empty())
+    return;
+  CircuitOp circuit = *circuits.begin();
+
+  // Walk all FModuleOps to potentially extract an OM class if the FModuleOp
+  // contains properties.
+  for (auto moduleOp : circuit.getOps<FModuleOp>())
+    extractClass(moduleOp);
```

Comment:
Did you consider parallelizing this (creating the `ClassOp`s and then working on their bodies in parallel)? This is likely tedious, but we will eventually need it.

Reply:
I think we can basically cut the extractClass implementation after the ClassOp is created, and call the second half in parallel for each ClassOp/ModuleOp. There will need to be some sharing of data computed before the ClassOp is created, but it can be done. I think I'll plan to leave this for a future optimization, since as you said, it's a little tedious but should be straightforward.